src/Renci.SshNet/ShellStream.cs

```diff
@@ -237,160 +231,188 @@ public override void SetLength(long value)
+                    }
+                }
+            }
         }
 
         /// <summary>
         /// Expects the expression specified by regular expression.
         /// </summary>
         /// <param name="regex">The regular expression to expect.</param>
         /// <returns>
-        /// The text available in the shell that contains all the text that ends with expected expression.
+        /// The text available in the shell up to and including the expected expression,
+        /// or <see langword="null"/> if the stream is closed without a match.
         /// </returns>
-        public string Expect(Regex regex)
+        public string? Expect(Regex regex)
         {
-            return Expect(regex, TimeSpan.Zero);
+            return Expect(regex, Timeout.InfiniteTimeSpan);
         }
 
         /// <summary>
         /// Expects the expression specified by regular expression.
         /// </summary>
         /// <param name="regex">The regular expression to expect.</param>
-        /// <param name="timeout">Time to wait for input.</param>
+        /// <param name="timeout">Time to wait for input. Must non-negative or equal to -1 millisecond (for infinite timeout).</param>
         /// <returns>
-        /// The text available in the shell that contains all the text that ends with expected expression,
-        /// or <see langword="null"/> if the specified time has elapsed.
+        /// The text available in the shell up to and including the expected expression,
+        /// or <see langword="null"/> if the specified timeout has elapsed or the stream
+        /// is closed without a match.
         /// </returns>
-        public string Expect(Regex regex, TimeSpan timeout)
+        /// <remarks>
+        /// If a TimeSpan representing -1 millisecond is specified for the <paramref name="timeout"/> parameter,
+        /// this method blocks indefinitely until either the regex matches the data in the buffer, or the stream
+        /// is closed (via disposal or via the underlying channel closing).
+        /// </remarks>
+        public string? Expect(Regex regex, TimeSpan timeout)
+        {
+            return ExpectRegex(timeout, [new ExpectAction(regex, s => { })]);
+        }
+
+        private string? ExpectRegex(TimeSpan timeout, ExpectAction[] expectActions)
         {
-            var result = string.Empty;
+            ValidateTimeout(timeout);
+
+            var timeoutTime = DateTime.Now.Add(timeout);
 
-            while (true)
+            lock (_sync)
             {
-                lock (_incoming)
+                while (true)
                 {
-                    if (_incoming.Count > 0)
-                    {
-                        result = _encoding.GetString(_incoming.ToArray(), 0, _incoming.Count);
-                    }
+                    AssertValid();
 
-                    var match = regex.Match(result);
+                    var bufferText = _encoding.GetString(_buffer, _head, _tail - _head);
```

Comment:
To integrate the expectSize, it would probably just be a case of restricting the search space here, e.g.

```c#
var expectHead = Math.Max(_head, _tail - _expectSize);
var bufferText = _encoding.GetString(_buffer, expectHead, _tail - expectHead);
```

The question is whether to put it as a parameter on the method or the constructor (I would probably prefer method)

Reply:
I also prefer a method parameter with the default value it will work for everyone without any breaking change

Reply:
I went with the parameter and named it `windowSize`, which felt to me a little more intuitive in meaning.

@jscarle, regretfully these changes (as they stand) reduce most of your recent contributions to be only in spirit rather than in code. I hope you consider the end result to be a net positive nonetheless. These recent changes wouldn't have happened without your motivating persistence.

Reply:
also on this line: unfortunately I think we still need to materialise the entire buffer as a string for `Expect(Regex)` in order to correctly deal with multi-byte encoding. (`Expect(string)` does not suffer from this because it works directly on the bytes). We can at least still use a reduced window for the regex match itself.

Reply:
In the end, what's important is that collaboratively we'll have managed to drastically improve Expect and fix bugs with ShellStream. These are issues that date back years now and made the previous implementation difficult to use, and in some cases impossible to use, in long running scenarios.

As for windowSize, I'd recommend against using the word window in this context. As we've been putting our hands in the plumbing, it may be obvious to us, however to someone consuming the API, it may not. I feel that it is too confusing and may lead people to misinterpret that as having some sort of relationship with the terminal window size.

In terms of performance, I'm glad that although my contribution has literally been obliterated by this PR, at least I was on par with your version in terms of execution time. My allocation improved, but your version is obviously much more efficient.

It would be a good idea to performance test Expect with a very large incoming buffer (in the several MB range) to ensure that there's no regression of the original performance issues that brought me to run a parallel queue to begin with.

Reply:
Right, `windowSize` is probably not a great name here. And I'd already been a bit confused by `expectSize`. I'll change it later. So far my best attempt is `rollingWindowSize`

Reply:
We could use one of these:
- match
- search
- inspect / inspection
- observation 
- peek
- frame
- analysis
- lookahead
- lookbehind

And/or a combination of above with Max, Size, Length, Range, or Bytes. ie: matchLength, peekSize.

Personally, I like matchLength, searchSize, inspect, and inspectSize. Though my most preferred is simply **inspect**. It's short, consice, and unambiguous.

Reply:
Thanks for the suggestions. I went with `lookback`